### PR TITLE
feat: add topic encoder

### PR DESCRIPTION
**Pull request by Rqnsom**
- function renaming in topic.rs
- add Display trait for the Tag struct

### CODE REVIEW

File: src/protocol/codecs/topic.rs [hunk 1]
```diff
@@ -81,6 +107,24 @@ impl TopicCodec {
 
         Ok(topics)
     }
+
+    /// Marshall topics to a byte stream.
+    fn marshall_topics(&mut self, topics: Vec<Topic>) -> BytesMut {
```
Comment (Rqnsom) [inline src/protocol/codecs/topic.rs:112]: We are aligning the function name to the original implementation here
https://github.com/algorand/go-algorand/blob/8a335bb2b10ea3cb6e9471526701bcaa4089b53d/network/msgOfInterest.go#L33

File: src/protocol/codecs/topic.rs [hunk 2]
```diff
@@ -40,6 +44,27 @@ impl MsgOfInterest {
...
+    #[allow(clippy::wrong_self_convention)]
+    fn to_topic(self) -> Topic {
+        let mut value = String::new();
+
+        if !self.tags.is_empty() {
+            for tag in self.tags {
+                value.push_str(&tag.to_string());
+                value.push(',');
+            }
+            value.pop().unwrap(); // Remove a trailing ','.
```
Comment (lwojtow) [inline src/protocol/codecs/topic.rs:58]: I think `join`, `map` and `collect` will do the same in more idiomatic way.

File: src/protocol/codecs/topic.rs [hunk 3]
```diff
@@ -81,6 +107,24 @@ impl TopicCodec {
 
         Ok(topics)
     }
+
+    /// Marshall topics to a byte stream.
+    fn marshall_topics(&mut self, topics: Vec<Topic>) -> BytesMut {
+        let mut raw_data = BytesMut::new();
+        let num_topics = topics.len() as u8;
```
Comment (lwojtow) [inline src/protocol/codecs/topic.rs:114]: I'm taking there is no chance for this to blow up? Maybe there should be some asserts here?

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:114]: raw_data can panic technically while doing these `put` operations, but we cannot control that (return Err if that happens),

but other than that, I don't see any need for error handling here. A vector of topics will always be a valid vector of topics, no?

File: src/protocol/codecs/topic.rs (hunk 2 above)
Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:58]: aha! let me try it 💪

File: src/protocol/codecs/topic.rs (hunk 3 above)
Comment (lwojtow) [reply src/protocol/codecs/topic.rs:114]: What I meant was this conversion `len() as u8`. Can len() be longer than 255?

File: src/protocol/codecs/topic.rs (hunk 2 above)
Comment (Rqnsom) [reply src/protocol/codecs/topic.rs]: It looks nicer! Thanks!
Done

File: src/protocol/codecs/topic.rs (hunk 3 above)
Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:111]: > Can len() be longer than 255?

Today, no. I think the largest value for the `len()` could be 5. I could leave a comment, so no one worries in the future

Comment (Rqnsom) [reply src/protocol/codecs/topic.rs:111]: Good catch!
So, [here](https://github.com/algorand/go-algorand/blob/8a335bb2b10ea3cb6e9471526701bcaa4089b53d/network/topics.go#L43):

```
// Topics is an array of type Topic
// The maximum number of topics allowed is 32
// Each topic key can be 64 characters long and cannot be size 0
type Topics []Topic
```
u8 is always okay for `num_topics` and `key_len`.

But this encoding/decoding won't work for the data part if the data size is longer than sizeof(u8).
So far, we are safe for this MsgOfInterest message.

I'll leave a TODO before lines 94 and 118 to check whether the variable length encoder/decoder is needed for future mesages.

File: src/protocol/codecs/tagmsg.rs [hunk 4]
```diff
@@ -30,6 +33,31 @@ pub enum Tag {
     VoteBundle,
 }
 
+impl Display for Tag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
```
Comment (niklaslong) [inline src/protocol/codecs/tagmsg.rs:41]: I'd suggest using the explicit names so a look at this match statement isn't necessary to understand logs; unless there's a reason for these shorthands?

**Review by niklaslong — APPROVED** [review]

File: src/protocol/codecs/tagmsg.rs (hunk 4 above)
Comment (Rqnsom) [reply src/protocol/codecs/tagmsg.rs:41]: > Maybe I could have implemented a normal (non Display) function to get these strings?

From our internal offline discussion: **Yes!**

The `Display` trait should be [used for user-facing output](https://doc.rust-lang.org/std/fmt/trait.Display.html):

> Display is similar to [Debug](https://doc.rust-lang.org/std/fmt/trait.Debug.html), but Display is for user-facing output, and so cannot be derived.

**Pull request merged**